binder/src/test/java/io/grpc/binder/BinderTransportSecurityTest.java

```diff
@@ -0,0 +1,235 @@
+ * limitations under the License.
+ */
+
+package io.grpc.binder;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.google.protobuf.Empty;
+
+import android.app.Application;
+import android.content.ComponentName;
+import android.content.Intent;
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.IBinder;
+import android.os.Looper;
+
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.protobuf.lite.ProtoLiteUtils;
+
+import androidx.lifecycle.LifecycleService;
+import androidx.test.core.app.ApplicationProvider;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.android.controller.ServiceController;
+
+import java.io.IOException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Executor;
+
+import io.grpc.MethodDescriptor;
+import io.grpc.Server;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerMethodDefinition;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.stub.ClientCalls;
+import io.grpc.stub.ServerCalls;
+
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricTestRunner.class)
+public final class BinderTransportSecurityTest {
+
+  private static final String SERVICE_NAME = "fake_service";
+  private static final String FULL_METHOD_NAME = "fake_service/fake_method";
+  private final Application context = ApplicationProvider.getApplicationContext();
+  private ServiceController<SomeService> controller;
+  private SomeService service;
+  private ManagedChannel channel;
+
+  @Before
+  public void setUp() {
+    controller = Robolectric.buildService(SomeService.class);
+    service = controller.create().get();
+
+    AndroidComponentAddress listenAddress = AndroidComponentAddress.forContext(service);
+    channel = BinderChannelBuilder.forAddress(listenAddress, context).build();
+    idleLoopers();
+  }
+
+  @After
+  public void tearDown() {
+    channel.shutdownNow();
+    controller.destroy();
+  }
+
+  @Test
+  public void testAsyncServerSecurityPolicy_failed_returnsFailureStatus() throws Exception {
+    ListenableFuture<Status> status = makeCall();
+    service.setSecurityPolicyStatusWhenReady(Status.ALREADY_EXISTS);
+    idleLoopers();
+
+    assertThat(status.get().getCode()).isEqualTo(Status.Code.ALREADY_EXISTS);
```

Comment:
Blocking the main thread risks deadlock. If this future is supposed to be done, can you call getDone() instead of blocking on get()?

Reply:
The future wasn't supposed to be done, because gRPC was running on its own ScheduledExecutorService, so I still had to wait until it finished running.

I switched to make all executors run on my tightly-controlled `Handler`. Now the future is always done. Switched to `Futures.getDone()`.